nearcache.go

```diff
@@ -0,0 +1,689 @@
+		ss:           ss,
+	}
+}
+
+func (m *nearCacheManager) GetOrCreateNearCache(name string, cfg nearcache.Config) *nearCache {
+	m.nearCachesMu.RLock()
+	nc, ok := m.nearCaches[name]
+	m.nearCachesMu.RUnlock()
+	if ok {
+		return nc
+	}
+	m.nearCachesMu.Lock()
+	nc, ok = m.nearCaches[name]
+	if !ok {
+		nc = newNearCache(&cfg, m.ss)
+		m.nearCaches[name] = nc
+	}
+	m.nearCachesMu.Unlock()
+	return nc
+}
+
+type nearCacheUpdateSemantic int8
+
+const (
+	nearCacheUpdateSemanticReadUpdate nearCacheUpdateSemantic = iota
+	nearCacheUpdateSemanticWriteUpdate
+)
+
+type nearCache struct {
+	store nearCacheRecordStore
+	cfg   *nearcache.Config
+}
+
+func newNearCache(cfg *nearcache.Config, ss *serialization.Service) *nearCache {
+	var rc nearCacheRecordValueConverter
+	var se nearCacheStorageEstimator
+	if cfg.InMemoryFormat == nearcache.InMemoryFormatBinary {
+		adapter := nearCacheDataStoreAdapter{ss: ss}
+		rc = adapter
+		se = adapter
+	} else {
+		adapter := nearCacheValueStoreAdapter{ss: ss}
+		rc = adapter
+		se = adapter
+	}
+	return &nearCache{
+		cfg:   cfg,
+		store: newNearCacheRecordStore(cfg, ss, rc, se),
+	}
+}
+
+func (nc *nearCache) Get(key interface{}) (interface{}, bool, error) {
+	_, ok := key.(serialization.Data)
+	if nc.cfg.SerializeKeys {
+		if !ok {
+			panic("key must be of type serialization.Data!")
+		}
+	} else if ok {
+		panic("key cannot be of type Data!")
+	}
+	return nc.store.Get(key)
+}
+
+func (nc *nearCache) Invalidate(key interface{}) {
+	nc.store.Invalidate(key)
+}
+
+func (nc nearCache) Size() int {
+	return nc.store.Size()
+}
+
+func (nc nearCache) Stats() nearcache.Stats {
+	return nc.store.Stats()
+}
+
+func (nc nearCache) InvalidationRequests() int64 {
+	return nc.store.InvalidationRequests()
+}
+
+func (nc *nearCache) TryReserveForUpdate(key interface{}, keyData serialization.Data, ups nearCacheUpdateSemantic) (int64, error) {
+	// eviction stuff will be implemented in another PR
+	// nearCacheRecordStore.doEviction(false);
+	return nc.store.TryReserveForUpdate(key, keyData, ups)
+}
+
+func (nc *nearCache) TryPublishReserved(key, value interface{}, reservationID int64) (interface{}, error) {
+	cached, err := nc.store.TryPublishReserved(key, value, reservationID, true)
+	if err != nil {
+		return nil, err
+	}
+	if cached != nil {
+		value = cached
+	}
+	return value, nil
+}
+
+const (
+	nearCacheRecordStoreTimeNotSet int64 = -1
+	pointerCostInBytes                   = (32 << uintptr(^uintptr(0)>>63)) >> 3
+	int32CostInBytes                     = 4
+	int64CostInBytes                     = 8
+	atomicValueCostInBytes               = 8
+	uuidCostInBytes                      = 16 // low uint64 + high uint64
+	nearCacheRecordNotReserved     int64 = -1
+	nearCacheRecordReadPermitted         = -2
+)
+
+type nearCacheRecordValueConverter interface {
+	ConvertValue(value interface{}) (interface{}, error)
+}
+
+type nearCacheStorageEstimator interface {
+	GetRecordStorageMemoryCost(rec *nearCacheRecord) int64
```

Comment:
Why not make key size estimation also a part of this?

Reply:
Java client implements `getKeyStorageMemoryCost` for both `NearCacheDataRecordStore` and `NearCacheObjectRecordStore`. If key is of type "serialization.Data" there is a cost but in our case, it never is I guess. Wdyt?

Reply:
I think getKeyStorageMemoryCost can directly return 0 because it is impossible to put a `serialization.Data` in the map.

Reply:
However, since we know the memory size of serialization.Data the cost it should be at least the cost of serialization.Data because string representation will take more space. Should we get the memory size of the serialization.Data? I think it is the most accurate one.

Reply:
I know but what should be the memory cost of the key in that case?

Reply:
I know. I think it is not accurate since we convert Data to string before storing in the `records` map. Note that, string representation of Data will take more space than Data. On the other hand, since we convert it to a string, and think it as a  `OBJECT` the cost should be 0. So should it be 0, `int64(bits.UintSize/8 + keyData.DataSize())` or more than `int64(bits.UintSize/8 + keyData.DataSize())` ?

Reply:
Why do you think the string representation takes more space that `Data` which is a byte slice?

Reply:
I thought it would print something like: 

`xff xde x12` but looks like it just stores the byte slice internally. Is this right?

Reply:
That's right.